compiler/rustc_lint/src/expect.rs

```diff
@@ -0,0 +1,381 @@
+use crate::builtin;
+use crate::context::{CheckLintNameResult, LintStore};
+use crate::late::unerased_lint_store;
+use crate::levels::{try_parse_reason_metadata, ParseLintReasonResult};
+use rustc_ast as ast;
+use rustc_ast::unwrap_or;
+use rustc_ast_pretty::pprust;
+use rustc_hir as hir;
+use rustc_hir::intravisit;
+use rustc_middle::hir::map::Map;
+use rustc_middle::ty::query::Providers;
+use rustc_middle::ty::TyCtxt;
+use rustc_session::lint::{Level, LintId};
+use rustc_session::Session;
+use rustc_span::symbol::{sym, Symbol};
+use rustc_span::{MultiSpan, Span};
+
+fn check_expect_lint(tcx: TyCtxt<'_>, _: ()) -> () {
+    if !tcx.sess.features_untracked().enabled(sym::lint_reasons) {
+        return;
+    }
+
+    let store = unerased_lint_store(tcx);
+    let krate = tcx.hir().krate();
+
+    let mut checker = LintExpectationChecker::new(tcx, tcx.sess, store);
+    checker.check_crate(krate);
+}
+
+/// This is used by the expectation check to define in which scope expectations
+/// count towards fulfilling the expectation.
+#[derive(Debug, Clone, Copy)]
+enum CheckScope {
+    /// The scope it limited to a `Span` only lint emissions within this span
+    /// can fulfill the expectation.
+    Span(Span),
+    /// All emissions in this crate can fulfill this emission. This is used for
+    /// crate expectation attributes.
+    CreateWide,
+}
+
+impl CheckScope {
+    fn includes_span(&self, emission: &MultiSpan) -> bool {
```

Comment:
Using spans is pretty hacky, and I suspect it might break because of things like macro expansion and stuff.

Reply:
This was one of the main problems with this task in general. Clippy and I suspect rustc as well emits almost all of its diagnostics via [`struct_span_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/context/trait.LintContext.html#method.struct_span_lint) which only provides us with very limited information. I asked on Zulip and there seems to be no real way to map from a [`Span`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html) to something like the `HirId`. This method could most likely support macros etc. by going through the span expansion data, but that seems performance intensive as it would also need to be done during `check_expect_query`.

Any other way of matching the emission span to the expectation would probably require a change to the way lints are emitted. Any suggestions are very welcome!

I like the suggestion you have [here](https://github.com/rust-lang/rust/pull/86024/files#r680366371) about tracking the expectations in the diagnostic emission. I'll first investigate possible solutions over there before going further into this. :upside_down_face: